Unable to import PnPjs modules correctly in NodeJS 16

### Category
- [ ] Enhancement
- [x] Bug
- [X] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [3.0.3]

Please specify what version(s) of SharePoint you are targeting: [SPO]

### Expected / Desired Behavior / Question

I am trying hard to use the v3 version of PnPjs, but the ESModules support is causing me a lot of trouble.
My scenario is a simple TypeScript project using [Azure Functions runtime 4 project](https://docs.microsoft.com/en-us/azure/azure-functions/functions-versions?tabs=in-process%2Cv4&pivots=programming-language-typescript), node 16.13.2 and TypeScript 4.7.0-dev.20220303.
Based on [TypeScript documentation](https://www.typescriptlang.org/docs/handbook/esm-node.html), I have this `tsconfig.json`:

```json
{
  "compilerOptions": {
    "module": "NodeNext",
    "target": "ES2017",
    "outDir": "dist",
    "rootDir": ".",
    "sourceMap": true,
    "strict": false
  }
}
```

I also set `"type": "module"` in my `package.json` file.

I have this simple `pnptest.mts` file:

```js
import { SPDefault } from "@pnp/nodejs";
import { spfi } from "@pnp/sp";
import "@pnp/sp/webs/index.js";
import { Configuration } from "@azure/msal-node";
// import "@pnp/core/util.js"

const config: Configuration = {
  auth: {
      authority: "https://login.microsoftonline.com/{tenant id or common}/",
      clientId: "{application (client) id}",
      clientSecret: "",
  },
};

const sp = spfi("https://318studios.sharepoint.com/sites/dev").using(SPDefault({
            msal: {
                config,
                scopes: [ 'https://{my tenant}.sharepoint.com/.default' ]
            },
        }));

const webData = await sp.web();
```

### Observed Behavior

`tsc` compilation works fine and produces the .mjs file as expected. But when I run the function I get this error:

```
[2022-03-03T16:38:50.927Z] Worker was unable to load function HelloHttp: 'Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/usr/app/azure/azure-functions/testnodepnpv4/node_modules/@pnp/core/util' imported from /usr/app/azure/azure-functions/testnodepnpv4/node_modules/@pnp/nodejs/sp-extensions/stream.js
[2022-03-03T16:38:50.927Z] Did you mean to import @pnp/core/util.js?'
```

If I uncomment the `import "@pnp/core/util.js"` I get the exact same error.

I guess I must be doing a dumb mistake (ESM is very new to me).  
And I noticed that in [your samples](https://pnp.github.io/pnpjs/concepts/authentication/#msal-nodejs) you don't speecify `index.js` in `import "@pnp/sp/webs/index.js"`, while I have to do it otherwise I have another error:

```
[2022-03-03T16:42:56.115Z] Worker was unable to load function HelloHttp: 'Error [ERR_UNSUPPORTED_DIR_IMPORT]: Directory import '/usr/app/azure/azure-functions/testnodepnpv4/node_modules/@pnp/sp/webs' is not supported resolving ES modules imported from /usr/app/azure/azure-functions/testnodepnpv4/dist/HelloHttp/pnptest.mjs
[2022-03-03T16:42:56.115Z] Did you mean to import @pnp/sp/webs/index.js?'
```

Thank you for your help

--- a/docs/getting-started.md
+++ b/docs/getting-started.md
@@ -168,6 +168,8 @@ export class SampleService {
 
 ## Getting started with NodeJS
 
+> Due to the way in which Node resolves ESM modules when you use selective imports in node you must include the `index.js` part of the path. Meaning an import like `import "@pnp/sp/webs"` in examples must be `import "@pnp/sp/webs/index.js"`. Root level imports such as `import { spfi } from "@pnp/sp"` remain correct. The samples in this section demonstrate this for their selective imports.
+
 ### Authentication
 
 To call the SharePoint APIs via MSAL you are required to use certificate authentication with your application. Fully covering certificates is outside the scope of these docs, but the following commands were used with openssl to create testing certs for the sample code below.
@@ -183,7 +185,7 @@ openssl rsa -in keytmp.pem -out key.pem -passin pass:HereIsMySuperPass
 
 ### Using @pnp/sp `spfi` factory interface in NodeJS
 
-> Version 3 of this library only supports ESModules. If for some reason you still require commonjs modules please check out [version 2](./v2/SPFx-on-premises/index.html).
+> Version 3 of this library only supports ESModules. If you still require commonjs modules please check out [version 2](./v2/SPFx-on-premises/index.html).
 
 The first step is to install the packages that will be needed. You can read more about what each package does starting on the [packages](packages.md) page.
 
@@ -196,7 +198,7 @@ Once these are installed you need to import them into your project, to communica
 ```TypeScript
 
 import { SPDefault } from "@pnp/nodejs";
-import "@pnp/sp/webs";
+import "@pnp/sp/webs/index.js";
 import { readFileSync } from 'fs';
 import { Configuration } from "@azure/msal-node";
 
@@ -242,7 +244,7 @@ Now we need to import what we'll need to call graph
 ```TypeScript
 import { graphfi } from "@pnp/graph";
 import { GraphDefault } from "@pnp/nodejs";
-import "@pnp/graph/users";
+import "@pnp/graph/users/index.js";
 
 function() {
     const graph = graphfi().using(GraphDefault({
@@ -289,7 +291,7 @@ setTimeout(async () => {
 
     const { spfi } = await import("@pnp/sp");
     const { SPDefault } = await import("@pnp/nodejs");
-    await import("@pnp/sp/webs");
+    await import("@pnp/sp/webs/index.js");
 
     const sp = spfi().using(SPDefault({
         baseUrl: settings.testing.sp.url,

--- a/docs/nodejs/behaviors.md
+++ b/docs/nodejs/behaviors.md
@@ -11,7 +11,7 @@ This behavior, for use in nodejs, provides basic fetch support through the `node
 ```TypeScript
 import { NodeFetch } from "@pnp/nodejs";
 
-import "@pnp/sp/webs";
+import "@pnp/sp/webs/index.js";
 
 const sp = spfi().using(NodeFetch());
 
@@ -21,7 +21,7 @@ await sp.webs();
 ```TypeScript
 import { NodeFetch } from "@pnp/nodejs";
 
-import "@pnp/sp/webs";
+import "@pnp/sp/webs/index.js";
 
 const sp = spfi().using(NodeFetch({ replace: false }));
 
@@ -35,7 +35,7 @@ This behavior makes fetch requests but will attempt to retry the request on cert
 ```TypeScript
 import { NodeFetchWithRetry } from "@pnp/nodejs";
 
-import "@pnp/sp/webs";
+import "@pnp/sp/webs/index.js";
 
 const sp = spfi().using(NodeFetchWithRetry());
 
@@ -47,7 +47,7 @@ You can also control how the behavior works through its props. The `replace` val
 ```TypeScript
 import { NodeFetchWithRetry } from "@pnp/nodejs";
 
-import "@pnp/sp/webs";
+import "@pnp/sp/webs/index.js";
 
 const sp = spfi().using(NodeFetchWithRetry({
     retries: 2,
@@ -77,7 +77,7 @@ You can use the baseUrl property to specify either v1.0 or beta - or one of the 
 ```TypeScript
 import { GraphDefault } from "@pnp/nodejs";
 import { graphfi } from "@pnp/graph";
-import "@pnp/graph/users";
+import "@pnp/graph/users/index.js";
 
 const graph = graphfi().using(GraphDefault({
     // use the German national graph endpoint
@@ -97,7 +97,7 @@ This behavior provides a thin wrapper around the `@azure/msal-node` library. The
 ```TypeScript
 import { MSAL } from "@pnp/nodejs";
 import { graphfi } from "@pnp/graph";
-import "@pnp/graph/users";
+import "@pnp/graph/users/index.js";
 
 const graph = graphfi().using(MSAL(config: { /* my msal config */ }, scopes: ["https://graph.microsoft.com/.default"]);
 
@@ -123,7 +123,7 @@ You can use the baseUrl property to specify the absolute site/web url to which q
 ```TypeScript
 import { SPDefault } from "@pnp/nodejs";
 
-import "@pnp/sp/webs";
+import "@pnp/sp/webs/index.js";
 
 const sp = spfi().using(SPDefault({
     msal: {
@@ -149,7 +149,7 @@ await sp.web();
 ```TypeScript
 import { StreamParse } from "@pnp/nodejs";
 
-import "@pnp/sp/webs";
+import "@pnp/sp/webs/index.js";
 
 const sp = spfi().using(StreamParse());
 

--- a/docs/nodejs/sp-extensions.md
+++ b/docs/nodejs/sp-extensions.md
@@ -35,11 +35,11 @@ const txt = await new Promise<string>((resolve) => {
 ```TypeScript
 import { SPDefault } from "@pnp/nodejs";
 import { spfi } from "@pnp/sp";
-import "@pnp/sp/webs";
-import "@pnp/sp/folders/web";
-import "@pnp/sp/folders/list";
-import "@pnp/sp/files/web";
-import "@pnp/sp/files/folder";
+import "@pnp/sp/webs/index.js";
+import "@pnp/sp/folders/web.js";
+import "@pnp/sp/folders/list.js";
+import "@pnp/sp/files/web.js";
+import "@pnp/sp/files/folder.js";
 import * as fs from "fs";
 
 const sp = spfi("https://something.com").using(SPDefault({
@@ -59,11 +59,11 @@ await files.addChunked(name, stream, null, true);
 ```TypeScript
 import { SPDefault } from "@pnp/nodejs";
 import { spfi } from "@pnp/sp";
-import "@pnp/sp/webs";
-import "@pnp/sp/folders/web";
-import "@pnp/sp/folders/list";
-import "@pnp/sp/files/web";
-import "@pnp/sp/files/folder";
+import "@pnp/sp/webs/index.js";
+import "@pnp/sp/folders/web.js";
+import "@pnp/sp/folders/list.js";
+import "@pnp/sp/files/web.js";
+import "@pnp/sp/files/folder.js";
 import * as fs from "fs";
 
 const sp = spfi("https://something.com").using(SPDefault({

--- a/packages/graph/outlook/types.ts
+++ b/packages/graph/outlook/types.ts
@@ -1,6 +1,6 @@
-import { _GraphQueryableCollection, _GraphQueryableInstance, graphInvokableFactory } from "@pnp/graph/graphqueryable";
+import { _GraphQueryableCollection, _GraphQueryableInstance, graphInvokableFactory } from "../graphqueryable.js";
 import { OutlookUser as IOutlookType, OutlookCategory as IOutlookCategoryType } from "@microsoft/microsoft-graph-types";
-import { defaultPath, deleteable, getById, IDeleteable, IGetById, IUpdateable, updateable } from "@pnp/graph/decorators";
+import { defaultPath, deleteable, getById, IDeleteable, IGetById, IUpdateable, updateable } from "../decorators.js";
 import { graphPost } from "@pnp/graph";
 import { body } from "@pnp/queryable";
 

--- a/packages/nodejs/sp-extensions/stream.ts
+++ b/packages/nodejs/sp-extensions/stream.ts
@@ -1,10 +1,9 @@
 import { headers } from "@pnp/queryable";
-import { File, Files, IFile, IFileAddResult, IFileInfo, IFiles, IFileUploadProgressData } from "@pnp/sp/files";
-import { spPost } from "@pnp/sp/operations";
+import { File, Files, IFile, IFileAddResult, IFileInfo, IFiles, IFileUploadProgressData } from "@pnp/sp/files/index.js";
+import { spPost, odataUrlFrom, escapeQueryStrValue } from "@pnp/sp";
 import { ReadStream } from "fs";
 import { PassThrough } from "stream";
 import { extendFactory, getGUID, isFunc } from "@pnp/core";
-import { odataUrlFrom, escapeQueryStrValue } from "@pnp/sp";
 import { StreamParse } from "../behaviors/stream-parse.js";
 
 export interface IResponseBodyStream {

